### ISSUE TITLE
Support TestCaseSource returning ValueTuple's for test cases

### ISSUE THREAD

**Issue by bhaeussermann** [nunit/nunit#3891]
It would be useful if `TestCaseSource` would support providing test cases as `System.ValueTuple` values instead of `object[]`.

Consider the following example. It currently throws the error "Not enough arguments provided, provide at least 2 arguments."

```
[TestFixture]
public class MyTest
{
    [TestCaseSource(nameof(TestCases))]
    public void Test(string name, int age)
    {
        // Testing code here.
    }

    public static (string name, int age)[] GetTestCases()
    {
        return new[]
        {
            ("John", 23), ("Sarah", 31)
        };
    }
}
```

The benefit of supporting the above is that it would allow the test case parameter types and number to be checked by the compiler in the test-case-source method. Though the example above is a bit contrived, this could be beneficial especially in the case of a more involved test-case-source method.

**Comment by rprouse** [comment]
This is an interesting idea. What do others think?

**Comment by stevenaw** [comment]
I like this idea, I think it it's a nice usability improvement. The example has both the tuple + method arg names matching, but would we want parameter passing be order-based in order to easily support unnamed valuetuples like below?

```csharp
public static (string, int)[] GetTestCases()
{
    return new[]
    {
        ("John", 23), ("Sarah", 31)
    };
}
```

**Comment by bhaeussermann** [comment]
@stevenaw Good point. If we make the parameters be matched based on the argument names, then that would allow for the orderings to differ between the tuples and the method parameters. This would be potentially confusing, and as such we should discourage the feature from being used this way.

Therefore, restricting the matching to be based on order will probably be the best approach. You can still use named tuples to aid readability.

**Comment by Thaina** [comment]
Are there any advantage of this approach over `TestCaseData` ? I think using TestCaseData might be more handy

**Comment by manfred-brands** [comment]
> Are there any advantage of this approach over `TestCaseData` ? I think using TestCaseData might be more handy

In TestCaseData all parameters are of type object. There is no compile time check.
Using specific types allows an analyzer to check this so developers get errors whilst developing vs runtime exceptions.

**Comment by CharliePoole** [comment]
We once (maybe still do) supported raw arrays as sources... both typed and object. This led to issues when the test method took a single argument of `int[]`, `object[]`, etc. We introduced `TestCaseData` and started discouraging use of the raw arrays.

I can imagine something similar happening here if the test method takes a tuple as an argument. I think use of tuples is an idea well worth considering, but the implementation will need to be tested against those kinds of corner cases.

**Comment by Thaina** [comment]
I just see we have `TestCaseData` generic

Are there anything like `TestCaseData.Create` that will auto infer generic for us?

**Comment by Thaina** [comment]
Just thinking another idea

Maybe we might have extension method for tuple

```cs
public static TestCaseData<T0,T1> AsTestCaseData<T0,T1>(in this (T0,T1) tuple) => new TestCaseData<T0,T1>();

//// usage

yield return (15,0.1f,"TestString").AsTestCaseData().SetArgDisplayNames("int 15","float 0.1f");
```
